src/logic/user-items-filter.ts

```diff
@@ -0,0 +1,150 @@
+import { Avatar, Outfit, Outfits } from '@dcl/schemas'
+import { BaseComponents, OnChainEmote, OnChainWearable } from '../types'
+import { splitUrnAndTokenId } from './utils'
+
+export type UserItemsFilter = {
+  filterNotOwnedItemsFromAvatars(avatar: Avatar[], owner: string): Promise<Avatar[]>
+  filterOutfitsWithoutCompleteOwnership(outfits: Outfits, owner: string): Promise<Outfits>
+}
+
+function isValidOutfit(outfit: Outfit, ownedWearables: OnChainWearable[]) {
+  const parsedWearables = outfit.wearables.map((wearable) => splitUrnAndTokenId(wearable))
+  return parsedWearables.every((wearable) => {
+    if (wearable.urn.includes('off-chain')) {
+      return true
+    }
+
+    const matchingOwnedWearable = ownedWearables.find(
+      (ownedWearable) =>
+        ownedWearable.urn === wearable.urn &&
+        (!wearable.tokenId || ownedWearable.individualData.some((itemData) => itemData.tokenId === wearable.tokenId))
+    )
+
+    return !!matchingOwnedWearable
+  })
+}
+
+function parseValidWearablesAndFilterInvalidOnes(
+  wearablesUrn: string[],
+  ownedWearables: OnChainWearable[],
+  shouldExtendWearables: boolean
+) {
+  return wearablesUrn
+    .map((wearable: string) => {
+      if (wearable.includes('base-avatars')) {
+        return wearable
+      }
+
+      const { urn, tokenId } = splitUrnAndTokenId(wearable)
+
+      const matchingOwnedWearable = ownedWearables.find(
+        (ownedWearable) =>
+          ownedWearable.urn === urn &&
+          (!tokenId || ownedWearable.individualData.find((itemData) => itemData.tokenId === tokenId))
+      )
+
+      if (!matchingOwnedWearable) {
+        return undefined
+      }
+
+      return shouldExtendWearables
+        ? `${matchingOwnedWearable.urn}:${tokenId ? tokenId : matchingOwnedWearable.individualData[0].tokenId}`
+        : matchingOwnedWearable.urn
+    })
+    .filter((wearable) => !!wearable) as string[]
```

Comment:
I would try to avoid the `as ..` it's very error-prone, I think in this case you could rewrite this as a for loop, I know it's a bit uglier but you would avoid typescript incorrectly infering types